pkg/scheduler/predicates.go

```diff
@@ -169,6 +169,119 @@ func PodFitsHost(pod api.Pod, existingPods []api.Pod, node string) (bool, error)
 	return pod.Spec.Host == node, nil
 }
 
+type NodeLabelChecker struct {
+	info     NodeInfo
+	labels   []string
+	presence bool
+}
+
+func NewNodeLabelPredicate(info NodeInfo, labels []string, presence bool) FitPredicate {
+	labelChecker := &NodeLabelChecker{
+		info:     info,
+		labels:   labels,
+		presence: presence,
+	}
+	return labelChecker.CheckNodeLabelPresence
+}
+
+// CheckNodeLabelPresence checks whether a particular label exists on a minion or not, regardless of its value
+// Consider the cases where the minions are placed in regions/zones/racks and these are identified by labels
+// In some cases, it is required that only minions that are part of ANY of the defined regions/zones/racks be selected
+//
+// Alternately, eliminating minions that have a certain label, regardless of value, is also useful
+// A minion may have a label with "retiring" as key and the date as the value
+// and it may be desirable to avoid scheduling new pods on this minion
+func (n *NodeLabelChecker) CheckNodeLabelPresence(pod api.Pod, existingPods []api.Pod, node string) (bool, error) {
+	var exists bool
+	minion, err := n.info.GetNodeInfo(node)
+	if err != nil {
+		return false, err
+	}
+	for _, label := range n.labels {
+		exists = labels.Set(minion.Labels).Has(label)
+		if (exists && !n.presence) || (!exists && n.presence) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+type ServiceAffinity struct {
+	podLister     PodLister
+	serviceLister ServiceLister
+	nodeInfo      NodeInfo
+	labels        []string
+}
+
+func NewServiceAffinityPredicate(podLister PodLister, serviceLister ServiceLister, nodeInfo NodeInfo, labels []string) FitPredicate {
```

Comment:
Use Case
We would like to ensure affinity for all pods within the same service. This can be achieved using the "MatchNodeSelector" predicate and by specifying the label value for pod.Spec.NodeSelector. However, for our use case, what region (as an example) a pod goes in is not relevant, and hence a user may not specify/select a region upfront. What is important is that every pod within the service goes into the same region, ensuring region affinity. Hence, this predicate allows the first pod for the service to be scheduled on a minion within any region (assuming a region is not spefcified in the pod.Spec.NodeSelector) and ensures that any subsequent pods within the service are hosted on minions within the same region as the first pod. Another use case is that you can have affinity at multiple levels. So, you could specify that you would like all pods for a service to be scheduled on minions within the same region and within the same zone (under the region). This is the reason to allow a list of labels to be specified in the ServiceAffinity struct.

Implementation: 
The levels (region/zone/rack/etc) at which the affinity is applied is governed by the ServiceAffinity struct "labels" field. By looking at the minion for already scheduled pods of the same service, we are eliminating the need to specify a region/zone upfront at the time of creating the pod (or replication controller).

Note:
There is nothing special about the terms -- region/zone/rack. These are just some terms used commonly within datacenters and hence have been used for ease of understanding. There is nothing preventing an admin from using city/building/room as labels for their infrastructure topological levels. 

Also, I will modify the code to first look at pod.Spec.NodeSelector labels instead of pod.Labels. This allows consistency with the other predicate and avoids confusion/conflcts.